### PR TITLE
Add state parameter check in oauth2 client

### PR DESCRIPTION
**Pull request by minlex**
Add support of the state parameter in the oauth2 client. This parameter is used for preventing cross-site request forgery.

### CODE REVIEW

File: socialregistration/clients/oauth.py [hunk 1]
```diff
@@ -324,6 +332,10 @@ def complete(self, GET):
             raise OAuthError(
                 _("Received error while obtaining access token from %s: %s") % (
                     self.access_token_url, GET['error']))
+
+        if self.state != GET['state']:
```
Comment (dbaxa) [inline socialregistration/clients/oauth.py]: This should be done using a constant time comparison function.

File: socialregistration/clients/oauth.py [hunk 2]
```diff
@@ -10,6 +11,8 @@
 import oauth2 as oauth
 import urllib
 import urlparse
+from random import SystemRandom
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: If we were to submit this upstream, it might be better to use `django.utils.crypto.get_random_string()`.

File: socialregistration/clients/oauth.py [hunk 3]
```diff
@@ -230,10 +233,15 @@ class OAuth2(Client):
     
     # Memoized user info fetched once an access token was obtained
     _user_info = None
-    
+
+    # The state for protection from CSRF
+    state = None
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: Should this be private?

File: socialregistration/clients/oauth.py [hunk 4]
```diff
@@ -243,14 +251,15 @@ def get_redirect_url(self, state='', **kwargs):
         Assemble the URL to where we'll be redirecting the user to to request
         permissions.
         """
+        if state != '':
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: When is the `state` parameter an empty string?

File: socialregistration/clients/oauth.py [hunk 5]
```diff
@@ -324,6 +333,10 @@ def complete(self, GET):
             raise OAuthError(
                 _("Received error while obtaining access token from %s: %s") % (
                     self.access_token_url, GET['error']))
+
+        if not constant_time_compare(self.state,GET['state']):
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: There should be a space after the comma here.

File: socialregistration/clients/oauth.py [hunk 6]
```diff
@@ -324,6 +333,10 @@ def complete(self, GET):
...
+            raise OAuthError("State does not match: %s" % GET['state'])
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: This error message seems kind of vague. Maybe it should be `"CSRF token missing or incorrect"`? (That's the string Django uses).

Also, this should be localized.

File: socialregistration/contrib/facebook/tests.py [hunk 7]
```diff
@@ -41,7 +41,7 @@ def get_facebook_data(self, *args, **kwargs):
     def callback(self, MockFacebook, MockRequest):
         MockRequest.side_effect = get_mock_func(self.get_callback_mock_response)
         MockFacebook.side_effect = get_mock_func(self.get_facebook_data)
-        response = self.client.get(self.get_callback_url(), {'code': 'abc'})
+        response = self.client.get(self.get_callback_url(), {'code': 'abc','state':self.state})
```
Comment (brodie) [inline socialregistration/contrib/facebook/tests.py]: There should be a space after the comma here.

File: socialregistration/tests.py [hunk 8]
```diff
@@ -249,12 +250,13 @@ class OAuth2Test(OAuthTest):
 
     def redirect(self):
         response = self.client.post(self.get_redirect_url())
+        self.state = re.findall('state=([\w\d]+)&',response['Location'])[0]
```
Comment (brodie) [inline socialregistration/tests.py]: There should be a space after the comma here.

File: socialregistration/tests.py [hunk 9]
```diff
@@ -249,12 +250,13 @@ class OAuth2Test(OAuthTest):
...
         return response
     
     @mock.patch('socialregistration.clients.oauth.OAuth2.request')
     def callback(self, MockRequest):
         MockRequest.side_effect = get_mock_func(self.get_callback_mock_response)
-        response = self.client.get(self.get_callback_url(), {'code': 'abc'})
+        response = self.client.get(self.get_callback_url(), {'code': 'abc','state':self.state})
```
Comment (brodie) [inline socialregistration/tests.py]: There should be a space after the comma here.

File: socialregistration/clients/oauth.py [hunk 10]
```diff
@@ -230,10 +233,15 @@ class OAuth2(Client):
...
-    
+
+    # The state for protection from CSRF
+    state = None
+
     def __init__(self, access_token=None):
         self._access_token = access_token
-    
+        self.state = ''.join(SystemRandom().choice(ascii_lowercase + digits)
+                             for _ in range(12))
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: `_` shadows an existing variable.

File: socialregistration/clients/oauth.py [hunk 11]
```diff
@@ -324,6 +328,10 @@ def complete(self, GET):
             raise OAuthError(
                 _("Received error while obtaining access token from %s: %s") % (
                     self.access_token_url, GET['error']))
+
+        if (not 'state' in GET) or (not constant_time_compare(self._state, GET['state'])):
```
Comment (brodie) [inline socialregistration/clients/oauth.py]: Style nitpick: you can use the `not in` operator here: (e.g., `'state' not in GET`). The parentheses would be superfluous at that point, I think (and they may already be).

File: readthedocs.txt [hunk 12]
```diff
@@ -1,2 +1,2 @@
-django>=1.3.1
+django>=1.4.0
```
Comment (brodie) [inline readthedocs.txt]: What's the purpose of this change?

**Comment by brodie** [comment]
Other than my very minor nitpicks, the change looks good to me.

The only other thing that I can think of is that it might be worth adding a test to assert that providing an invalid state doesn't work.

File: readthedocs.txt (hunk 12 above)
Comment (minlex) [inline readthedocs.txt]: Version 1.3.1 does not have get_random_string function. Should I stick to version 1.3.1?

**Comment by dbaxa** [comment]
Can we send this change upstream? or does social auth already contain a fix?

**Comment by minlex** [comment]
I think the original project is long dead.

**Comment by dbaxa** [comment]
Right.
